lib/resty/dns/resolver.lua

```diff
@@ -300,109 +308,18 @@ local function _build_request(qname, id, no_recurse, opts)
     }
 end
 
+local function parse_section(answers, section, buf, start_pos, size, should_skip)
+    local pos = start_pos
 
-local function parse_response(buf, id)
-    local n = #buf
-    if n < 12 then
-        return nil, 'truncated';
-    end
-
-    -- header layout: ident flags nqs nan nns nar
-
-    local ident_hi = byte(buf, 1)
-    local ident_lo = byte(buf, 2)
-    local ans_id = lshift(ident_hi, 8) + ident_lo
-
-    -- print("id: ", id, ", ans id: ", ans_id)
-
-    if ans_id ~= id then
-        -- identifier mismatch and throw it away
-        log(DEBUG, "id mismatch in the DNS reply: ", ans_id, " ~= ", id)
-        return nil, "id mismatch"
-    end
-
-    local flags_hi = byte(buf, 3)
-    local flags_lo = byte(buf, 4)
-    local flags = lshift(flags_hi, 8) + flags_lo
-
-    -- print(format("flags: 0x%x", flags))
-
-    if band(flags, 0x8000) == 0 then
-        return nil, format("bad QR flag in the DNS response")
-    end
-
-    if band(flags, 0x200) ~= 0 then
-        return nil, "truncated"
-    end
-
-    local code = band(flags, 0xf)
-
-    -- print(format("code: %d", code))
-
-    local nqs_hi = byte(buf, 5)
-    local nqs_lo = byte(buf, 6)
-    local nqs = lshift(nqs_hi, 8) + nqs_lo
-
-    -- print("nqs: ", nqs)
-
-    if nqs ~= 1 then
-        return nil, format("bad number of questions in DNS response: %d", nqs)
-    end
-
-    local nan_hi = byte(buf, 7)
-    local nan_lo = byte(buf, 8)
-    local nan = lshift(nan_hi, 8) + nan_lo
-
-    -- print("nan: ", nan)
-
-    -- skip the question part
-
-    local ans_qname, pos = _decode_name(buf, 13)
-    if not ans_qname then
-        return nil, pos
-    end
-
-    -- print("qname in reply: ", ans_qname)
-
-    -- print("question: ", sub(buf, 13, pos))
-
-    if pos + 3 + nan * 12 > n then
-        -- print(format("%d > %d", pos + 3 + nan * 12, n))
-        return nil, 'truncated';
-    end
-
-    -- question section layout: qname qtype(2) qclass(2)
-
-    local type_hi = byte(buf, pos)
-    local type_lo = byte(buf, pos + 1)
-    local ans_type = lshift(type_hi, 8) + type_lo
-
-    -- print("ans qtype: ", ans_type)
-
-    local class_hi = byte(buf, pos + 2)
-    local class_lo = byte(buf, pos + 3)
-    local qclass = lshift(class_hi, 8) + class_lo
-
-    -- print("ans qclass: ", qclass)
-
-    if qclass ~= 1 then
-        return nil, format("unknown query class %d in DNS response", qclass)
-    end
-
-    pos = pos + 4
-
-    local answers = {}
-
-    if code ~= 0 then
-        answers.errcode = code
-        answers.errstr = resolver_errstrs[code] or "unknown"
-    end
-
-    for i = 1, nan do
+    for i = 1, size do
         -- print(format("ans %d: qtype:%d qclass:%d", i, qtype, qclass))
-
         local ans = {}
-        insert(answers, ans)
+
+        if not should_skip then
+            insert(answers, ans)
+        end
+
+        ans.section = section
 
```

Comment:
The Lua variables `type_hi` and `type_lo` below are not declared as `local` and thus are examples of global variable abuses.